grpc/nodejs/rules.bzl

```diff
@@ -14,116 +14,96 @@
+    all_outputs = []
+    is_windows_host = ctx.configuration.host_path_separator == ";"
+    for target in ctx.attr.deps:
+        args = ctx.actions.args()
+        # Output and Plugin path
+        args.add(_as_path(ctx.executable._protoc_gen_ts.path, is_windows_host), format = "--plugin=protoc-gen-ts=%s")
+        args.add(paths.join(ctx.bin_dir.path, paths.dirname(ctx.build_file_path)), format = "--ts_out=%s")
+
+        # Set in descriptors
+        descriptor_sets_paths = [desc.path for desc in target[ProtoInfo].transitive_descriptor_sets.to_list()]
+        args.add_joined("--descriptor_set_in", descriptor_sets_paths, join_with = ctx.configuration.host_path_separator)
+
+        # Direct sources
+        args.add_all(_imported_protos(target))
+
+        if not len(ctx.outputs.outs):
+            outputs = _get_outputs(target, ctx)
+            all_outputs.extend(outputs)
+        else:
+            outputs = ctx.outputs.outs
+            all_outputs = ctx.outputs.outs
+
+        ctx.actions.run(
+            inputs = depset(target[ProtoInfo].direct_sources, transitive = [target[ProtoInfo].transitive_descriptor_sets]),
+            tools = [ctx.executable._protoc_gen_ts],
+            executable = ctx.executable._protoc,
+            outputs = outputs,
+            arguments = [args],
+            progress_message = "Generating Protocol Buffers for Typescript %s" % ctx.label,
+            env = {"BAZEL_BINDIR": ctx.bin_dir.path},
         )
-    },
-    implementation = _proto_sources
-)
 
-unpack_proto_archive = rule(
+    return [
+        DefaultInfo(files = depset(all_outputs))
+    ]
+
+
+ts_grpc_compile = rule(
+    implementation = _node_protoc,
     attrs = {
         "deps": attr.label_list(
+            doc = "List of proto_library targets.",
             providers = [ProtoInfo],
+            mandatory = True
         ),
-        "grpc_deps": attr.label_list(
-             providers = [ProtoInfo],
-         ),
-        "archive": attr.label(
-            allow_single_file = True
+        "outs": attr.output_list(),
+        "_protoc": attr.label(
+            cfg = "exec",
+            executable = True,
+            default = "@com_google_protobuf//:protoc"
         ),
-    },
-    implementation = _unpack_proto_archive
+        "_protoc_gen_ts": attr.label(
+            cfg = "exec",
+            executable = True,
+            default = "@npm//protoc-gen-ts/bin:protoc-gen-ts"
+        )
+   }
 )
-
-
-def ts_grpc_compile(
-        name,
-        deps,
-        grpc_deps):
-    proto_sources_name = "{}__do_not_reference_1".format(name)
-    genrule_name = "{}__do_not_reference_2".format(name)
-    proto_sources(
-        name = proto_sources_name,
-        deps = deps
-    )
-    native.genrule(
-        name = genrule_name,
-        outs = [
-            "{}.tar.gz".format(genrule_name),
-        ],
-        # The below command performs a protoc compilation of our proto files into typescript and javascript. Line by line:
```

Comment:
Notably, we got rid of all this custom implementation in a `genrule` and converted it to a proper rule, because one of our dependencies requires access to `runfiles`, which a genrule cannot use.